core/src/services/supabase/backend.rs

```diff
@@ -0,0 +1,342 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::fmt::Debug;
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use http::HeaderValue;
+use http::StatusCode;
+use log::debug;
+
+use super::core::*;
+use super::error::parse_error;
+use super::writer::*;
+use crate::ops::*;
+use crate::raw::*;
+use crate::*;
+
+/// Supabase service
+///
+/// # Capabilities
+///
+/// - [x] read
+/// - [x] write
+/// - [ ] copy
+/// - [ ] list
+/// - [ ] scan
+/// - [ ] presign
+/// - [ ] blocking
+///
+/// # Configuration
+///
+/// - `root`: Set the work dir for backend.
+/// - `bucket`: Set the container name for backend.
+/// - `endpoint`: Set the endpoint for backend.
+/// - `service_key`: Set the secret key for the backend, this is prioritized over the client key
```

Comment:
We don't need to introduce the two different key here. It's users own decision. How about just use `key`? Like supabase's example:

```js
void main() async {
  final supabase = SupabaseClient('supabaseUrl', 'supabaseKey');

  // Create file `example.txt` and upload it in `public` bucket
  final file = File('example.txt');
  file.writeAsStringSync('File content');
  final storageResponse = await supabase
      .storage
      .from('public')
      .upload('example.txt', file);
}
```

Reply:
One more thing to talk about. Should be expose API like
```Rust
builder.anon_key();
builder.service_key();
```
and call `public/auth` API separately based on that.

OR

we only provide
```Rust
builder.auth_key()
```
and let the server return unauthorized when user uses a anon_key to access a unpublic resource.

Reply:
I prefer to keep consist with supabase's SDK, call it `key` and return error if users trying to visit unpublic resource.

Reply:
We can mention the `service_key` and `anon_key` in the docs of `fn key()` to let users know the difference.

Reply:
> We can mention the `service_key` and `anon_key` in the docs of `fn key()` to let users know the difference.

So, in our implementation. We should try twice on accessing the resource since the urls for `public` and `authorized` are different.

It would look like
```shell
if supabase_get_object_info_public().failed() {
    supabase_get_object_info_auth()
}
```

Reply:
Set `AUTHORIZATION` header if we have a `key`, otherwise leave it empty. Users should make the decision, we just return the server's response back,

Reply:
> Hi, supabase doesn't allow calling `write` without `authorization`. Please try `read` a public bucket.

Oops, I got your point. You are right.

Another case, suppose we are calling read() on a path, the underlying operation would send a get request to a url.
The authorized url and the public url is different, to be more specific, `/authenticated/bucket/wildcard` and /`public/bucket/wildcard`.
Which function should we call? Since we are not bookeeping the key is anon_key or service_key.

IMO, we should call the `public` one first, if failed, then call the `authenticated` one.

What do you think?

Reply:
> The authorized url and the public url is different, to be more specific, `/authenticated/bucket/wildcard` and /`public/bucket/wildcard`.

I guess we should call `authenticated` if `key` has been set, otherwise, call `public`. 

Seems we have confused the `public` bucket and supabase's `public` API key. From storage's API spec, `public` API doesn't need any auth header.

---

> IMO, we should call the `public` one first, if failed, then call the `authenticated` one.

No, we never do this in OpenDAL.

Reply:
What if the user loads an `anon_key`, i.e., filling the `key: Option<T>` field. But then it initiates a read, which should we call, the public one or the authenticated one, since we do not know the key is anon or service.

In the SDK provided by supabase, its `read()` can have another argument like `public/authenticated`, but this is not the case for us in OpenDAL since our read operation is started by Operator without that kinds of arguments.

Reply:
We don't need to support that case. It's easier for users to build two operator with different keys.

Reply:
> We don't need to support that case. It's easier for users to build two operator with different keys.

Maybe we have some misunderstanding here.
Take the following code segment as an example
```Rust

let b = services::Supabase::default();
b.key("some_anon_key");

let op = Operator::new(b);

let _ = op.read("some_key");
```

Underlying this read would be the `Accessor::read()`.

At the end of the day, we have to call some functions to actually read the data from Supabase. Currently we have
`supabase_get_object_auth()` and `supabase_get_object_public()`.

If we choose the auth one, i.e.: `supabase_get_object_auth()`
- For those does not set the key / set with anon_key: the read will fail even if the bucket is public.

If we choose the public one, i.e.: `supabase_get_object_public()`
- For those set with service_key, they cannot access the nonpublic bucket even they have their authorized key.

In general, the real problem is, the `Operator`'s read does not specify whether we should call the `supabase_get_object_auth()` or `supabase_get_object_public()`, and if we always choose one of them to be called in the backend. The problem described earlier will happen.

Reply:
Leave  to users to decide. They know what bucket they are visiting. It's not opendal's job.

Reply:
> Leave to users to decide. They know what bucket they are visiting. It's not opendal's job.
> 
> We just need to make our behaviour clear.

The problem is that the user cannot decide which url to choose. Supabase uses different url to access secret buckets and public buckets, and we only provide a unified interface of `op.read()`

The url for authorized bucket is `/object/authorized/{bucket_name}/{wild_card}`, for the public bucket is `/object/public/{bucket_name}/{wild_card}`. We have to request to either one of them or both of them, how can user decide which one to call through `op.read()`?

Reply:
If they want to call public bucket, don't pass a key; otherwise, pass a key.

Reply:
> If they want to call public bucket, don't pass a key; otherwise, pass a key.

But this draws the previous problem that, without a key they cannot write to a public bucket.

So you mean, if they want to do so, they should have to construct two operators, one with anon_key to write and another one with no key to read?

I think it would be fine, we can polish that later on.